internal/metadata/metadata.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2016-2023, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metadata
+
+import (
+	"encoding/json"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+)
+
+// The underlying value of a metadata blob.
+type Data struct {
+	m map[string]json.RawMessage
+}
+
+func New(data []byte) (*Data, error) {
+	m := map[string]json.RawMessage{}
+	if data != nil {
+		err := json.Unmarshal(data, &m)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &Data{m}, nil
+}
+
+func (d *Data) Marshal() []byte {
+	if d == nil {
+		d = &Data{m: make(map[string]json.RawMessage)}
+	}
+	bytes, err := json.MarshalIndent(d.m, "", "    ")
+	// `d.m` is a `map[string]json.RawMessage`. `json.MarshalIndent` errors only when
+	// it is asked to serialize an unmarshalable type (complex, function or channel)
+	// or a cyclic data structure. Because `string` and `json.RawMessage` are
+	// trivially marshallable and cannot contain cycles, all values of `d.m` can be
+	// marshaled without error.
+	//
+	// See https://pkg.go.dev/encoding/json#Marshal for details.
+	contract.AssertNoErrorf(err, "internal: failed to marshal metadata")
+	return bytes
+}
+
+// Set a piece of metadata to a value.
+//
+// Set errors only if value fails to serialize.
+func Set(d *Data, key string, value any) error {
```

Comment:
Would this be more natural as `func (*Data) Set(key, value)`?

Reply:
I'm trying to create a separation between consumers of `Data` (the tfbridge package) and holders/passers of metadata (end providers). To name and pass around metadata, you need to import `github.com/pulumi/pulumi-terraform-bridge/v3/pkg/metadata`. To manipulate metadata, you need to import `github.com/pulumi/pulumi-terraform-bridge/v3/unstable/metadata`. 

If we make `Set` a method (as opposed to a function), it will be usable from `github.com/pulumi/pulumi-terraform-bridge/v3/pkg/metadata.Data` and not require the `unstable` import. 

Note: I used to have `Set` in `internal/metadata`, but `pf` can't use that, so I changed `internal` to `unstable`.

Reply:
https://github.com/pulumi/pulumi-terraform-bridge/blob/a7f49a7d5354a98b924aa3ffc62dafcdcdbd2850/unstable/doc.go#L15-L20